### DISCUSSION TITLE
Audio Tools Github Library

### DISCUSSION THREAD

**Discussion by HobmanX** [pschatzmann/arduino-audio-tools#705] [General]
Is there anythis wrong with Github Audio Tools library Zip file. I tried to download Zip and only get 17818 kb and it does not work.
I still have copy of older Feb 2023 Zip and it works OK......????

**Comment by pschatzmann** [comment]
What's the error that you are getting ?
I never install it this way...

**Reply by HobmanX** [reply]
No error message, library installs OK via add Zip library Arduino 2.0.3. It asks if I want to overwrite old library, I say Yes, and all installs.
I open sketch and Upload to ESP 32 Pico D4 kit, all look OK, but no sound. I am using pins 23,19,and 22, left justified with PT8211 DAC.
If I re-install copy of older Feb23 Zip by add zip library all works without any changes to sketch etc.
The only thing I notice is new Zip file is 17818 KB  and old Feb zip is  54747 KB

**Comment by pschatzmann** [comment]
If you changed any settings in the AudioConfig.h then this gets overwritten.
Apart from this I am not aware of any changes that could cause this.
I suggest that you set the log level to info and double check the pins...

I moved all the pictures and audio which are not used by any sketch to a separate project, therefore it is smaller now!

**Reply by HobmanX** [reply]
I have tried various sketches and various pin setups and run the logger. It is still doing the same...... The Old Feb Library works no matter what pins for I2S and no matter what sketch.
I have included your basic tone sketch here and you will see I have commented out/in some line to try various pins and setups. This all works with old Feb library but not on the one downloaded today. I either get no sound or "Bad" digital noise in speaker with todays lib.

What is the size of the new library Zip File ??......... Is there a problem with my download ??..... My Zip is 17818 kb ??

//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

`/**
 * @file streams-generator-i2s.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-i2s/README.md 
 * @copyright GPLv3
 */
 
#include "AudioTools.h"

uint16_t sample_rate=44100;
uint8_t channels = 2;                                      // The stream will have 2 channels 
SineWaveGenerator<int16_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
I2SStream out; 
StreamCopy copier(out, sound);                             // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.sample_rate = sample_rate; 
  config.channels = channels;
  config.bits_per_sample = 16;

// for external I2S below...pins used ESP32 PICO KIT.....
   config.pin_bck = 4;       //   4 ... 23 ... 14
   config.pin_ws = 5;        //   5  ... 19 ... 15
   config.pin_data = 19;      //  19 ... 22 ... 22

  // For external I2S below .... pins used on RP2040 PICO ...
   //config.pin_bck = 20;
  // config.pin_ws = 21;
  // config.pin_data = 22;

  // config.i2s_format = I2S_STD_FORMAT;
   config.i2s_format = I2S_LSB_FORMAT;   // for PT8211 DAC ext

  out.begin(config);

  // Setup sine wave
  sineWave.begin(channels, sample_rate, N_B4);
  Serial.println("started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}

`
Below is logs with both libraries
//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Startup and logs below...........

good Old Feb library

ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13220
ho 0 tail 12 room 4
load:0x40080400,len:3028
entry 0x400805e4
starting I2S...
started...

mCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops

//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

march 23 library .... not working ??

ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13220
ho 0 tail 12 room 4
load:0x40080400,len:3028
entry 0x400805e4
starting I2S...
started...


logger.............
 - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops

//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

**Comment by pschatzmann** [comment]
Did you try to change the logger to Warning?
I have decreased the number of buffers a from 10 to 6
I think that I also changed the I2S_USE_APLL from false to true and I2S_AUTO_CLEAR from false to true
to fix some strange repeating noise on silence.

So I suggest that you try to play with these I2S settings.
config. buffer_count = 10;
config.auto_clear = false;
config.use_apll = false;
I suggest to change only one parameter at a time or to compare with values in the log of your version that is working.

**Reply by HobmanX** [reply]
If I change to Warning ..... nothing is listed in the terminal
I made the changes you suggest........ no difference......no sound

Can you confirm the size of the Zip download file..... I may have a problem with Norton virus checker or proxy etc...

**Reply by pschatzmann** [reply]
As I told you, I can't help with zip, because that's not the way I work.
I am also confused, with what settings are you getting noise and with what settings not output at all ?
Have you compared the pins in the log with the physical pins ?

**Reply by HobmanX** [reply]
""As I told you, I can't help with zip, because that's not the way I work.""...... What other way is there to update...??.  I have always used the Zip method for years...It is in the menu of the Arduino IDE. .....  under Sketch ..... Include Library ......Add Zip Library......find it on PC.....click on it and it installs. Because there is already an old one installed it asks "do you want to overwrite".....I say Yes... and it tells me "Sucessfully installed..  First I need to download the actual Zip library file from your Github.....click the Code Tab ...... Click Download File.... and the file downloads OK.    I have never done anything else no matter whose library I was updating.
""I am also confused, with what settings are you getting noise and with what settings not output at all ?""...The tone sketch posted earlier is the one I am trying.....when I tried it earlier this morning I was not getting any sound then later in afternoon I get crackling as in digital noise.....using the setting as in the post and also the minor changes and pin changes you suggest.
""Have you compared the pins in the log with the physical pins ?"" .... How do I do  that. Any of the logs I see in the terminal is just pages and pages scrolling up the screen, similar to the bits I have posted.....

**Comment by pschatzmann** [comment]
Just before you see the [I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops lines,
you should see the __relevant configuration information__. Just start and as soon as you see the StreamCopy you can stop. 

Maybe you need to press the reboot button to see everything because the Serial interface is taking too long of you power it up. You could also add a long delay to the setup. Unfortunatly you  pasted the information that is not important at all...

You can compare this with the version that is working to see any differencies...

E.g. I am getting
19:55:12.428 -> [I] I2SConfig.h : 106 - rx/tx mode: TX_MODE
19:55:12.428 -> [I] I2SConfig.h : 107 - port_no: 0
19:55:12.428 -> [I] I2SConfig.h : 108 - is_master: Master
19:55:12.428 -> [I] I2SConfig.h : 109 - sample rate: 48000
19:55:12.428 -> [I] I2SConfig.h : 110 - bits per sample: 16
19:55:12.428 -> [I] I2SConfig.h : 111 - number of channels: 2
19:55:12.428 -> [I] I2SConfig.h : 112 - i2s_format: I2S_STD_FORMAT
19:55:12.428 -> [I] I2SConfig.h : 114 - auto_clear: true
19:55:12.460 -> [I] I2SConfig.h : 116 - use_apll: true
19:55:12.460 -> [I] I2SConfig.h : 121 - buffer_count:6
19:55:12.460 -> [I] I2SConfig.h : 122 - buffer_size:512
19:55:12.460 -> [I] I2SConfig.h : 126 - pin_bck: 27
19:55:12.460 -> [I] I2SConfig.h : 128 - pin_ws: 25
19:55:12.460 -> [I] I2SConfig.h : 130 - pin_data: 26

**Reply by HobmanX** [reply]
This is what I get with the same basic example sketch as posted above run for each report
The only differencr is updating the Zip Library file from Feb23 copy to the one downloaded today
//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Not working 23rd March 23 library
I hear only hiss and crackles... no broken up tone

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13220
ho 0 tail 12 room 4
load:0x40080400,len:3028
entry 0x400805e4
starting I2S...
[I] I2SConfig.h : 106 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 107 - port_no: 0
[I] I2SConfig.h : 108 - is_master: Master
[I] I2SConfig.h : 109 - sample rate: 44100
[I] I2SConfig.h : 110 - bits per sample: 16
[I] I2SConfig.h : 111 - number of channels: 2
[I] I2SConfig.h : 112 - i2s_format: I2S_LSB_FORMAT
[I] I2SConfig.h : 114 - auto_clear: false
[I] I2SConfig.h : 116 - use_apll: true
[I] I2SConfig.h : 121 - buffer_count:10
[I] I2SConfig.h : 122 - buffer_size:512
[I] I2SConfig.h : 126 - pin_bck: 4
[I] I2SConfig.h : 128 - pin_ws: 5
[I] I2SConfig.h : 130 - pin_data: 19
[I] SoundGenerator.h : 173 - SineWaveGenerator::begin(channels=2, sample_rate=44100, frequency=493.88)
[I] SoundGenerator.h : 159 - bool audio_tools::SineWaveGenerator<T>::begin() [with T = short int]
[I] SoundGenerator.h : 46 - bits_per_sample: 16
[I] SoundGenerator.h : 47 - channels: 2
[I] SoundGenerator.h : 48 - sample_rate: 44100
[I] SoundGenerator.h : 200 - setFrequency: 493.88
[I] SoundGenerator.h : 201 - active: true
started...
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops

//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Working but crackling due to serial print and logger info ......
I hear a mix of tone and crackles.....would be OK in not logging...I assume
This is with Feb 2023 library

s Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13220
ho 0 tail 12 room 4
load:0x40080400,len:3028
entry 0x400805e4
starting I2S...
[I] I2SConfig.h : 104 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 105 - port_no: 0
[I] I2SConfig.h : 106 - is_master: Master
[I] I2SConfig.h : 107 - sample rate: 44100
[I] I2SConfig.h : 108 - bits per sample: 16
[I] I2SConfig.h : 109 - number of channels: 2
[I] I2SConfig.h : 110 - i2s_format: I2S_LSB_FORMAT
[I] I2SConfig.h : 112 - auto_clear:0
[I] I2SConfig.h : 114 - use_apll: true
[I] I2SConfig.h : 119 - buffer_count:10
[I] I2SConfig.h : 120 - buffer_size:512
[I] I2SConfig.h : 124 - pin_bck: 4
[I] I2SConfig.h : 126 - pin_ws: 5
[I] I2SConfig.h : 128 - pin_data: 19
[I] SoundGenerator.h : 173 - SineWaveGenerator::begin(channels=2, sample_rate=44100, frequency=493)
[I] SoundGenerator.h : 159 - bool audio_tools::SineWaveGenerator<T>::begin() [with T = short int]
[I] SoundGenerator.h : 46 - bits_per_sample: 16
[I] SoundGenerator.h : 47 - channels: 2
[I] SoundGenerator.h : 48 - sample_rate: 44100
[I] SoundGenerator.h : 200 - setFrequency: 493
[I] SoundGenerator.h : 201 - active: true

started...
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops

**Reply by pschatzmann** [reply]
Hmm, this is identical. After some more checks I noticed that I have changed the obsolete i2s_format. The Documentation states that the old I2S_COMM_FORMAT_I2S |I2S_COMM_FORMAT_I2S_LSB is now I2S_COMM_FORMAT_STAND_MSB which is very confusing. 
So can you try with 

config.i2s_format = I2S_MSB_FORMAT

I also decided to roll back this change and keep the old conventions for the time being...

**Reply by HobmanX** [reply]
Yes, that got the basic tone example working, but tried my own sketch with same line, it still not working,  I will look at it again tomorrow

**Reply by HobmanX** [reply]
I got a quick check on my own sketch................
It plays OK with the old Feb Zip Library, but with March 23rd library no sound at all no crackling
 Sketch is resample and input mixer sketch
Below is output from logger and serial println()
Serial.println() added at, start setup, setup end, running loop, running playlist
 Output shows .... Sketch only gets as far as setup end ..... no more output on terminal
So sketch is not running loop or playlist.......code shown below at end................

//   March 23rd Zip Library .... No sound out
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13220
ho 0 tail 12 room 4
load:0x40080400,len:3028
entry 0x400805e4
running setup
[I] AudioTypes.h : 63 - sample_rate: 44100
[I] AudioTypes.h : 64 - channels: 1
[I] AudioTypes.h : 65 - bits_per_sample: 16
[I] Resample.h : 106 - setStepSize: 0.687500
[I] AudioTypes.h : 63 - sample_rate: 44100
[I] AudioTypes.h : 64 - channels: 1
[I] AudioTypes.h : 65 - bits_per_sample: 16
[I] Resample.h : 106 - setStepSize: 1.000000
[I] I2SConfig.h : 106 - rx/tx mode: TX_MODE
[I] I2SConfig.h : 107 - port_no: 0
[I] I2SConfig.h : 108 - is_master: Master
[I] I2SConfig.h : 109 - sample rate: 48000
[I] I2SConfig.h : 110 - bits per sample: 16
[I] I2SConfig.h : 111 - number of channels: 1
[I] I2SConfig.h : 112 - i2s_format: I2S_MSB_FORMAT
[I] I2SConfig.h : 114 - auto_clear: true
[I] I2SConfig.h : 116 - use_apll: true
[I] I2SConfig.h : 121 - buffer_count:10
[I] I2SConfig.h : 122 - buffer_size:512
[I] I2SConfig.h : 126 - pin_bck: 4
[I] I2SConfig.h : 128 - pin_ws: 5
[I] I2SConfig.h : 130 - pin_data: 19
[I] SoundGenerator.h : 530 - bool audio_tools::GeneratorFromArray<T>::begin() [with T = short int]
[I] SoundGenerator.h : 46 - bits_per_sample: 16
[I] SoundGenerator.h : 47 - channels: 1
[I] SoundGenerator.h : 48 - sample_rate: 48000
[I] SoundGenerator.h : 530 - bool audio_tools::GeneratorFromArray<T>::begin() [with T = short int]
[I] SoundGenerator.h : 46 - bits_per_sample: 16
[I] SoundGenerator.h : 47 - channels: 1
[I] SoundGenerator.h : 48 - sample_rate: 48000
[I] Resample.h : 106 - setStepSize: 0.687500
setup end

//cccccccccccccccccccccccccccccccccccccccccccccccc

//this code is last line of setup and loop and playlist

Serial.println("setup end");
}  //setup end

// Arduino loop - copy sound to out 

void loop() {
  copier1.copy();
  //copier2.copy();
Serial.println("running loop");
// soundplay();

   playlist(); 
}

void playlist(void) {
  b = millis();
 
if (listp >= 9) listp=0;

if ((b - a) >= 1000) {
Serial.println("running playlist");
resample2.setStepSize(NoteArray[listp]*freqf);   
 a = millis();
 listp++;
}

}

//ccccccccccccccccccccccccccccccccc

**Reply by pschatzmann** [reply]
Can you provide a simple comprehensive sketch where I can reproduce your issue ?

**Reply by HobmanX** [reply]
This is the a basic example sketch reading an audio 1 cycle array of 256 sample, resampling and mixing various pitches against a low freq continuous tone to see if it is plays a steady tone with no beats. You will remember back in January #556 I had a long thread regarding this problem and was not sorted out at that time, and I had been away from it all for a while so did not keep track.  I looked at it again some time ago and that may have been when I updated to the Feb 23 Zip Library, but at that time I did notice that it played at that time ver steady and no beats also in January thre was a problem with the wave amplitude having sudden changes.  When I use the February Zip library now I find it plays very well and constant amplitude and no beats although I am now using the PT8211 for better quality audio.
The Zip file I downloaded yesterday gives me no output at all with this same sketch in the same Arduino IDE 2.0.3.
This is the sketch attached
[PSchatx2.txt](https://github.com/pschatzmann/arduino-audio-tools/files/11064999/PSchatx2.txt)

**Reply by HobmanX** [reply]
I see there were some changes on the GitHub this morning relating to I2S New....??? so I downloaded another Zip and updated. 
I tried sketch with the same settings and still same ....No sound out....logger shows it stops at end of setup and does nor run Loop.
I thought I would try the internal DAC audio analog out pins 25 and 26..... Also just the same No Audio out for todays download Mar25th or my last download Mar23rd.   Feb plays good audio out internal DAC pins 25, 26. and as I stated in my last post for external DAC the problems of "beating" has gone and the amplitude is now constant and sound clear and steady....maybe better than ext DAC.  So what ever happened between  my January tests and February tests has sorted out the sound.  But I am only checking with this sketch sinewaves for now ... would need to test with real sounds and square waves to see for harmonic distortion etc.
So still problem with the current Zip files on the Github........
By the way..... the beating is still on when I try maximillian sketches as far as I know.....so your library seems better now...except for lib problem now....

**Reply by pschatzmann** [reply]
IDF has changed the I2S implementation in 5.0. Unfortunatley this is not available in Arduino yet.
I could find the cause of your issue by setting the log level to debug: It seems that all writes to a buffer were failing.
I could trace down the issue to the new resize() method in the ringbuffer.

With the latest commit it should work again...

**Reply by HobmanX** [reply]
Ok, thanks, I have just logged in and see you posted..... but I came on to say I tried maximillian TwoTones sketch and it worked with  yesterday Zip file with I2S_MSB_FORMAT for the external DAC, but not work with I2S_LSB_FORMAT (just noise). And it worked with internal DAC at default config.     I will try out your latest post now....thanks

**Reply by HobmanX** [reply]
Yes, the sketch with the resample and mixing works OK with the new Mar25th Zip File......... but for the External PT8211 it needs I2S_LSB_FORMAT ........ 
With regards to mixing and setWeight........I have totally confused myself.
Previously, I had been using this as sort of amplitude..... with a value between 0 and 1  as I could not see any difference in amplitude if above 1.   I had it at 0.25 and 1.0 in the sketch posted earlier......????
When I was trying the sketch now I could not hear the second tone so set the louder 1.0 down to 0.8 and then the sound stopped making me think it was not working.  The volume does not change no matter where it is.... It seems to work on the difference between the 2 values for how much of the second tone is heard. I now see in the popup help that comes in IDE 2.0.3 that it uses values  between 0 and 100....??
Have you any examples how to use volume control...??
  mixer.add(resample1);     
  mixer.add(resample2);     
  mixer.setWeight(0, 40);     // 0.25  
  mixer.setWeight(1, 40);     //  1.0

**Reply by pschatzmann** [reply]
I undid the change for the i2s format, so it's working as before.
Just replace your current output to i2s to the volume control and specify the volume control to go to I2S....
https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control

**Reply by HobmanX** [reply]
Yes, it all seems to be working well now.....but didn't get the volume working as your example link........
  // set initial volume
  volume.begin(config); // we need to provide the bits_per_sample and channels
  volume.setVolume(0.3);
I could not figure out why volume.begin(config);  and where I would add the bits per sample and channels...?
It seems to only affect the I2S output..... I want to control the input streams to the mixer....vary the amplitude on the fly of each wave being mixed. So I figured out something that seems to work..... would welcome your views and comments... I have attached the working bit of the sketch, the sample array and top bit are same as before ..... so will paste together....
[PSchatx2x.txt](https://github.com/pschatzmann/arduino-audio-tools/files/11070131/PSchatx2x.txt)

**Reply by pschatzmann** [reply]
That should work as well, however it is easier if you just dynamically adjust the weights of the mixer instead. 
This requires no additional complexity...

For further details consult the [documentation](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_input_mixer.html#af958d7d7ce83c9577da34c091d216aa3)

>> I could not figure out why volume.begin(config); and where I would add the bits per sample and channels...?

I do not understand your question, you currently already pass it a configured [I2SConfig](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_i2_s_config.html) which is a subclass of AudioBaseInfo.

**Reply by HobmanX** [reply]
Regards mixer  doing similar to volume...??? 
For me it doesn't work as a volume of the individual streams but seems to work as a relative volume between them, probably similar to a stereo balance control, although I cannot see each stream separately as output....
If I set......
  mixer.setWeight(0, 0);     
  mixer.setWeight(1, 1);
I can hear and see on audacity stream 1 on both I2S outputs L&R and I don't hear stream 0....so I assume stream 0 is muted...?

 If I set ....
  mixer.setWeight(0, 1);     
  mixer.setWeight(1, 0);
I can hear and see on audacity stream 0 on both I2S outputs L&R and I don't hear stream 1....so I assume stream 1 is muted...?

 also it I set .......
  mixer.setWeight(0, 0);     
  mixer.setWeight(1, 80);

OR

  mixer.setWeight(0, 80);     
  mixer.setWeight(1, 0);

I hear same as before.....no change in volume and just hear the one stream on L&R channels....... values don't matter

If I set.......
  mixer.setWeight(0, 1);     
  mixer.setWeight(1, 1);
I hear both streams now mixing loudly together with same volume amplitude as before.

If I set.......
  mixer.setWeight(0, 10);     
  mixer.setWeight(1, 10);
OR
If I set.......
  mixer.setWeight(0, 50);     
  mixer.setWeight(1, 50);

I still hear both streams mixing loudly with same volume. It doesn't matter the value between 1 to 100 if both numbers are same then both mix loudly.

If I set.......
  mixer.setWeight(0, 10);     
  mixer.setWeight(1, 90);
I hear stream 0 lower volume than stream 1 and both mixing
OR
If I set.......
  mixer.setWeight(0, 90);     
  mixer.setWeight(1, 10);
I hear stream 1 lower volume than stream 0 and both mixing
It doesn't matter what numbers what seems to matter is the ratio between them.
so I assume for 10 and 90 it means 0.1 for one and 0.9 volume for other
 similarily.....

  mixer.setWeight(0, 9);     
  mixer.setWeight(1, 1);
is same as 10 and 90
or 5 and 45, or 4 and 36 etc etc one is 9 times the other in volume assumed from what I hear.

So I think volumeStream on the mixer inputs is needed for individual stream control and volumeStream on output I2S is needed for Overall volume controll without upsetting the mixer settings..... like a mixing desk....

The resample and mix sketch ESP32 is OK for playing Sine waves and triangle waves that are smooth...it seems to be able to cope with that, but for arbitrary real sound waves and square waves it can create a lot of harmonic distortion or "aliases" etc...I would need to play with it a bit more.

Regards you say I should  "read the documentation", I have looked at this various times but would need a book on how to read the documentation.... Is there any good web links that describes all this stuff....?

**Comment by pschatzmann** [comment]
The class documentation recommends that you should use a weight of 100 as max value. This gives you quite a big resolution of values to choose from. If you decide to use 1 you just specify that you want to switch it on or off.  

The system internally calculates the ratio from the sum of all values. If all values are the same the output it mixed with identical weights. So in the end 900/100 or 90/10 or 9/1 give all the same result!
